geniza/annotations/migrations/0003_associate_related_footnotes.py

```diff
@@ -0,0 +1,118 @@
+        raise Resolver404("Not a document manifest URL")
+    return resolve_match.kwargs["pk"]
+
+
+def associate_related_footnotes(apps, schema_editor):
+    """
+    Data migration to associate all existing annotations with footnotes by foreign key,
+    instead of using source and manifest URIs in the content field.
+    """
+
+    User = apps.get_model("auth", "User")
+    Annotation = apps.get_model("annotations", "Annotation")
+    ContentType = apps.get_model("contenttypes", "ContentType")
+    Document = apps.get_model("corpus", "Document")
+    Footnote = apps.get_model("footnotes", "Footnote")
+    Source = apps.get_model("footnotes", "Source")
+    LogEntry = apps.get_model("admin", "LogEntry")
+
+    # stats for report
+    bad_manifest_uris = set()
+    failed_annotations = set()
+    created_footnotes = 0
+
+    # much of this code adapted from signal handlers and static methods
+    document_contenttype = ContentType.objects.get_for_model(Document)
+
+    annotations = Annotation.objects.all()
+    for annotation in annotations:
+        # resolve URIs to Source and Document IDs (without using static methods)
+        manifest_uri = (
+            annotation.content.get("target", {})
+            .get("source", {})
+            .get("partOf", {})
+            .get("id")
+        )
+        source_uri = annotation.content.get("dc:source")
+        source_id = int(source_uri.split("/")[-2])
+        try:
+            document_id = document_id_from_manifest_uri(manifest_uri)
+        except Resolver404:
+            bad_manifest_uris.add(manifest_uri)
+            failed_annotations.add(annotation.pk)
+            continue
+        try:
+            # try to get a DIGITAL_EDITION footnote for this source and document
+            footnote = Footnote.objects.get(
+                doc_relation=["X"],  # DIGITAL_EDITION at this point has a value of "X"
+                source__pk=source_id,
+                content_type=document_contenttype,
+                object_id=document_id,
+            )
+        except Footnote.DoesNotExist:
+            # create the DIGITAL_EDITION footnote
+            source = Source.objects.get(pk=source_id)
+            footnote = Footnote.objects.create(
+                source=source,
+                doc_relation=["X"],
+                object_id=document_id,
+                content_type=document_contenttype,
+            )
+            LogEntry.objects.log_action(
+                user_id=User.objects.get(username=settings.SCRIPT_USERNAME).id,
+                content_type_id=ContentType.objects.get_for_model(Footnote).pk,
+                object_id=footnote.pk,
+                object_repr=str(footnote),
+                action_flag=ADDITION,
+                change_message=f"Footnote automatically created via annotation migration.",
+            )
+            created_footnotes += 1
+
+        # instantiate the FK relation
+        annotation.footnote = footnote
```

Comment:
I think this migration should also remove the manifest and source uris from the content of the annotation — otherwise we will risk them getting out of sync